polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1343,6 +1343,130 @@ class RaiseSetNDRange
+      rewriter.create<sycl::SYCLHostSetCaptured>(captureOp->getLoc(), lambdaObj,
+                                                 rewriter.getI64IntegerAttr(0),
+                                                 capturedVal);
+    }
+
+    // All other captures are unique assignments to GEPs with two constant
+    // indices `[0, <capture #>]` to the lambda object.
+    for (auto *user : lambdaObj.getUsers()) {
+      auto gep = dyn_cast<LLVM::GEPOp>(user);
+      if (!gep || !gep.getElemType().has_value() ||
+          gep.getElemType().value() != lambdaClassTy ||
+          !gep.getDynamicIndices().empty())
+        continue;
+
+      auto indices = gep.getRawConstantIndices();
+      if (indices.size() != 2 || indices[0] != 0)
+        continue;
+      unsigned captureIdx = indices[1];
+
+      auto [captureOp, capturedVal] = getUniqueAssignment(gep);
+      if (!captureOp)
+        continue;
+
+      capturedVal = tryToBroaden(capturedVal, captureTypes[captureIdx]);
+      rewriter.setInsertionPointAfter(captureOp);
+      rewriter.create<sycl::SYCLHostSetCaptured>(
+          captureOp->getLoc(), lambdaObj,
+          rewriter.getI64IntegerAttr(captureIdx), capturedVal);
+    }
+
+    // Finally erase the annotation op.
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+private:
+  /// If there is exactly one store or memcpy to \p ptr, return it and the
+  /// stored value/copied-from pointer; otherwise return `nullptr` and an emtpy
+  /// value.
+  static std::tuple<Operation *, Value>
+  getUniqueAssignment(Value ptr, bool allowMemcpy = true) {
+    Operation *op = nullptr;
+    Value value;
+    for (auto *user : ptr.getUsers())
+      if (auto v = TypeSwitch<Operation *, Value>(user)
+                       .Case<LLVM::StoreOp>([&](auto st) {
+                         return st.getAddr() == ptr ? st.getValue() : Value();
+                       })
+                       .Case<LLVM::MemcpyOp>([&](auto mc) {
+                         return mc.getDst() == ptr ? mc.getSrc() : Value();
+                       })
+                       .Default(Value())) {
+        if (op)
+          // Assignment is not unique.
+          return {nullptr, Value()};
+
+        op = user;
+        value = v;
+      }
+
+    return {op, value};
+  }
+
+  // Use the \p expected type as domain knowledge to try to broaden a \p stored
+  // value to an entity of interest (e.g. an accessor).
+  Value tryToBroaden(Value stored, Type expected) const {
+    if (isClassType(expected, accessorTypeTag.getTypeName())) {
+      // The getelementpointer[0, <capture #>] we have matched earlier is is not
+      // addressing the entire accessor, but rather the first member in the
+      // accessor class (think: getelementpointer[0, <capture #>, 0...]).
+      // Instead of the value loaded from that address, we return the address
+      // itself, which points to the accessor.
+      auto load = cast<LLVM::LoadOp>(stored.getDefiningOp());
```

Comment:
If the defining operation of the Value `stored` is not a Load this would assert. How are we sure that the def. operation will always be a load ?

Reply:
My initial thinking was that it would be ok to err-out when the pattern I observed for accessors doesn't match, but after pondering more about it, I agree and changed the code to match the common structure more precisely, and fall-back to not touching the argument if it doesn't.